### PR TITLE
Implement claim metadata deletion on tenant deletion

### PR DESCRIPTION
**Pull request by sumedhe**
### Proposed changes in this pull request

Implement deletion of claim metadata when a Tenant gets deleted

Fixes: #2931

### CODE REVIEW

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java [hunk 1]
```diff
@@ -151,4 +151,14 @@ void removeExternalClaim(String externalClaimDialectURI, String externalClaimURI
     default void removeClaimMappingAttributes(int tenantId, String userstoreDomain) throws ClaimMetadataException {
 
     }
+
+    /**
+     * Remove all claims of a given tenant.
+     *
+     * @param tenantId Id of the tenant
+     * @throws ClaimMetadataException
```
Comment (sachiniWettasinghe) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:159]: Complete the javadoc comment.

Comment (sachiniWettasinghe) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:159]: Check every other instance and fix this.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java [hunk 2]
```diff
@@ -366,6 +366,19 @@ public void removeClaimMappingAttributes(int tenantId, String userstoreDomain) t
...
+    /**
+     * Remove all claims of a given tenant.
+     *
+     * @param tenantId Id of the tenant
+     * @throws ClaimMetadataException
+     */
+    @Override
+    public void removeAllClaims(int tenantId) throws ClaimMetadataException {
+
+        // The relevant external claim deletions are handled by the DB through ON DELETE CASCADE
```
Comment (sachiniWettasinghe) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java:378]: Add proper formatting. The comment should end with a "."

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDialectDAO.java [hunk 3]
```diff
@@ -136,4 +136,30 @@ public void removeClaimDialect(ClaimDialect claimDialect, int tenantId) throws C
...
+     * @param tenantId Id of the tenant
+     * @throws ClaimMetadataException
+     */
+    public void removeAllClaimDialects(int tenantId) throws ClaimMetadataException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+        PreparedStatement prepStmt = null;
+
+        String query = SQLConstants.REMOVE_CLAIM_DIALECTS_BY_TENANT_ID;
+        try {
```
Comment (sachiniWettasinghe) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDialectDAO.java:152]: Any reason here to not to use try with resource option?

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package org.wso2.carbon.identity.claim.metadata.mgt.listener;
```
Comment (sachiniWettasinghe) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java:1]: Add liscence header

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.wso2.carbon.identity.claim.metadata.mgt.exception.ClaimMetadataException;
+import org.wso2.carbon.identity.claim.metadata.mgt.internal.IdentityClaimManagementServiceDataHolder;
+import org.wso2.carbon.stratos.common.beans.TenantInfoBean;
+import org.wso2.carbon.stratos.common.exception.StratosException;
+import org.wso2.carbon.stratos.common.listeners.TenantMgtListener;
+
+public class ClaimMetadataTenantMgtListener implements TenantMgtListener {
```
Comment (sachiniWettasinghe) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java:9]: Add the class description comment.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java (hunk 1 above)
Comment (sumedhe) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:159]: Fixed in #3312

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java (hunk 2 above)
Comment (sumedhe) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java:378]: Fixed in #3312

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDialectDAO.java (hunk 3 above)
Comment (sumedhe) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDialectDAO.java:152]: Fixed in #3312

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java (hunk 4 above)
Comment (sumedhe) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java:1]: Fixed in #3312

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java (hunk 5 above)
Comment (sumedhe) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/listener/ClaimMetadataTenantMgtListener.java:9]: Fixed in #3312

**Pull request merged**